src/node.h

```diff
@@ -1473,6 +1473,9 @@ class NODE_EXTERN AsyncResource {
   Environment* env_;
   v8::Global<v8::Object> resource_;
   async_context async_context_;
+#if defined(NODE_USE_NATIVE_ALS) && NODE_USE_NATIVE_ALS
+  v8::Global<v8::Value> context_frame_;
```

Comment:
Any idea how to make this _not_ a semver-major? I would ideally like to be able to backport this as much as possible. 🤔

Reply:
Sorry, no idea. While this here can be solved by marking this semver major the NAPI part is even harder as NAPI should be ABI compatible even across major versions.

`AsyncResource` has already a member `async_context` and that one should carry all needed data for Napi and node API.

Reply:
I wanted to look at this PR for a few weeks now but got caught with other work and did not get to it. Now I had the chance and came across this discussion.
Since `node.h` is supposed to be as stable as possible within a major version, like @Flarna mentioned for `napi` even across multiple major versions probably the only way would be to have this added in a `std::unordered_map<size_t, v8::Global<v8::Value>>` in `class Environment`.
`env_` is accessible from the AsyncResource and this would not impact the ABI compatibility as far as I can tell. At least `sizeof(AsyncResource)` stays the same.
I tried some things with it and pushed [a commit](https://github.com/dygabo/node/commit/02e7bfefe257135de5b9dcadb7df90f4c55c7f46) on a [branch copy](https://github.com/dygabo/node/tree/gabo/qard-review-async-local-storage-rewrite) on my node fork. 
@Qard If it fits the needs of this PR feel free to take that commit over on your branch. By using an unordered_map the lookup is in constant time (O(1)) on average and the insert in amortized constant so it should by a fast operation.